File: 031_next_permutation/Solution.py
class Solution:
    def nextPermutation(self, nums: list) -> None:
        arr_len = len(nums)
        if arr_len <= 1: return

        i, j, k = arr_len - 2, arr_len - 1, arr_len - 1

        while i >= 0 and nums[i] >= nums[j]:
            i -= 1
            j -= 1

        if i >= 0:
            while nums[i] >= nums[k]: k -= 1
            nums[i], nums[k] = nums[k], nums[i]

        i, j = j, arr_len - 1
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1


if __name__ == '__main__':
    # nums = [1, 2, 3, 4, 5, 6]
    nums = [1, 1]
    solution = Solution()
    solution.nextPermutation(nums)
    print(nums)

File: 031_next_permutation/Solution.java
class Solution {
    public void nextPermutation(int[] nums) {
        int arrLen = nums.length;
        if (arrLen <= 1) return;

        int i = arrLen - 2, j = arrLen - 1, k = arrLen - 1;

        while (i >= 0 && nums[i] >= nums[j]) {
            i--; j--;
        }

        if (i >= 0) {
            while (nums[i] >= nums[k]) k--;
            swap(i, k, nums);
        }

        for (i = j, j = arrLen - 1; i < j; i++, j--) {
            swap(i, j, nums);
        }
    }

    public void swap(int i, int j, int[] nums) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1, 2, 3};
        Solution solution = new Solution();
        solution.nextPermutation(nums);
        for (int num : nums)
            System.out.print(num + " ");
    }
}
